src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/PoolingDataSource.cs

```diff
@@ -45,19 +83,224 @@ internal sealed class PoolingDataSource : SqlDataSource
+        /// </summary>
+        /// <param name="connector">The connector to be checked.</param>
+        /// <returns>Returns true if the connector is live and unexpired, otherwise returns false.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private bool CheckIdleConnector([NotNullWhen(true)] SqlConnector? connector)
+        {
+            if (connector is null)
+                return false;
+
+            // Only decrement when the connector has a value.
+            Interlocked.Decrement(ref _idleCount);
+
+            // An connector could be broken because of a keepalive that occurred while it was
+            // idling in the pool
+            // TODO: Consider removing the pool from the keepalive code. The following branch is simply irrelevant
+            // if keepalive isn't turned on.
+            if (connector.IsBroken)
+            {
+                CloseConnector(connector);
+                return false;
+            }
+
+            /* TODO: enforce connection lifetime
+            if (_connectionLifetime != TimeSpan.Zero && DateTime.UtcNow > connector.OpenTimestamp + _connectionLifetime)
+            {
+                LogMessages.ConnectionExceededMaximumLifetime(_logger, _connectionLifetime, connector.Id);
+                CloseConnector(connector);
+                return false;
+            }
+
+            // The connector directly references the data source type mapper into the connector, to protect it against changes by a concurrent
+            // ReloadTypes. We update them here before returning the connector from the pool.
+            Debug.Assert(SerializerOptions is not null);
+            Debug.Assert(DatabaseInfo is not null);
+            connector.SerializerOptions = SerializerOptions;
+            connector.DatabaseInfo = DatabaseInfo;
+
+            Debug.Assert(connector.State == ConnectorState.Ready,
+                $"Got idle connector but {nameof(connector.State)} is {connector.State}");
+            Debug.Assert(connector.CommandsInFlightCount == 0,
+                $"Got idle connector but {nameof(connector.CommandsInFlightCount)} is {connector.CommandsInFlightCount}");
+            Debug.Assert(connector.MultiplexAsyncWritingLock == 0,
+                $"Got idle connector but {nameof(connector.MultiplexAsyncWritingLock)} is 1");
+            */
+            return true;
+        }
+
+        /// <summary>
+        /// Closes the provided connector and adjust pool state accordingly.
+        /// </summary>
+        /// <param name="connector">The connector to be closed.</param>
+        private void CloseConnector(SqlConnector connector)
+        {
+            try
+            {
+                connector.Close();
+            }
+            catch
+            {
+                //TODO: LogMessages.ExceptionWhenClosingPhysicalConnection(_logger, connector.Id, exception);
+            }
+
+            var i = 0;
+            for (; i < MaxPoolSize; i++)
```

Comment:
Please use `{}`

In my opinion, seeing `for (; ...)` invokes a 'wat' response. Though I see the reason for doing it, you want to access `i` outside of the `for` block. Here's a couple options for less 'wat' response:
* `int i; for (i = 0; ...)`
* `int i = 0; while (i < MaxPoolSize) { ... i++; }`